### PR TITLE
Feature/data loader

### PR DESCRIPTION
**Pull request by moon1ite**
# Pull Request
레파지토리에 기여해주셔서 감사드립니다.

해당 PR을 제출하기 전에 아래 사항이 완료되었는지 확인 부탁드립니다:
- [x] 작성한 코드가 어떤 에러나 경고없이 빌드가 되었나요?
- [x] 충분한 테스트를 수행하셨나요?

## 1. 해당 PR은 어떤 내용인가요?

- minor bug fix
  - tokenizer 쪽에 `<eos> <bos> <pad>` 를 reserve하지 않았어서 다시 학습하여 올립니다
  - 불필요한 코드는 제거하였습니다
- main contributions
  - raw text를 읽어서 위에서 학습한 tokenizer 를 통해 encoding, padding 
  - 향후 training에 사용할 data loader 

## 2. PR과 관련된 이슈가 있나요?

#1

### CODE REVIEW

File: src/dataset.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+import torch
+from omegaconf import OmegaConf
+from tokenizers import SentencePieceBPETokenizer
+from torch.utils.data import Dataset
+
+root_dir = Path(__file__).parent.parent
+model_config_dir = root_dir / "configs" / "model"
+tokenizer_config_dir = root_dir / "configs" / "tokenizer"
+tokenizer_dir = root_dir / "tokenizer"
```
Comment (ssaru) [inline src/dataset.py]: 이 내용을 global로 빼기보다, `WMT14Dataset`의 생성자(`__init__`) 파라미터로 받아서 처리하는 것은 어떨까요?

File: src/dataset.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    """
+
+    def __init__(
+        self, langpair: str, source_lines: List[str], target_lines: List[str]
+    ) -> None:
+        super().__init__()
+        if langpair in ["de-en", "en-de", "deen", "ende"]:
+            self.tokenizer_config_path = (
+                tokenizer_config_dir / "sentencepiece_bpe_wmt14_deen.yaml"
+            )
```
Comment (ssaru) [inline src/dataset.py]: 이거는 suggenstion은 아니구요.
저는 보통 이런 방식도 즐겨 사용하긴합니다만, 요런건 어떻게 생각하시나요?!

이 방식 외에도 `REGISTRY`방식을 사용하는 것도 괜찮아보이구요.
https://detectron2.readthedocs.io/tutorials/datasets.html#register-a-dataset

```suggestion
class WMT14Dataset(Dataset):
    """Dataset to train and test transformer

    Attributes:
        langpair: language pair to translate
        source_lines: list of text in source language
        target_lines: list of text in target language
    """
    LANG: List[str] = ["de-en", "en-de", "deen", "ende"]
    def __init__(
        self, langpair: str, source_lines: List[str], target_lines: List[str]
    ) -> None:
        super().__init__()
        if langpair in self.LANG:
            self.tokenizer_config_path = (
                tokenizer_config_dir / "sentencepiece_bpe_wmt14_deen.yaml"
            )
```

File: src/dataset.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        #  elif langpair in ["en-fr", "fr-en", "enfr", "fren"]:
+        #      self.tokenizer_config_path = (
+        #          tokenizer_config_dir / "sentencepiece_bpe_wmt14_enfr.yaml"
+        #      )
+        else:
+            raise NotImplementedError(
+                f'{langpair} is not supported. Hephaestus project aims to reproduce "Attention is all you need", where transformer is tested on de-en and en-fr.'
+            )
+        self.model_config = OmegaConf.load(model_config_dir / "transformers.yaml")
+        self.tokenizer_config = OmegaConf.load(self.tokenizer_config_path)
```
Comment (ssaru) [inline src/dataset.py]: 이 부분도 `get_config`이라는 함수로 묶을 수 있을 것 같다는 느낌이 듭니다.
https://github.com/HephaestusProject/pytorch-Sencnn/blob/master/train.py

File: src/load_dataset.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+from omegaconf import OmegaConf
+from pytorch_lightning import LightningDataModule
+from torch.utils.data import DataLoader
+
+from .dataset import WMT14Dataset
+from .utils import read_lines
+
+root_dir = Path(__file__).parent.parent
+dataset_config_dir = root_dir / "configs" / "dataset"
+model_config_dir = root_dir / "configs" / "model"
```
Comment (ssaru) [inline src/load_dataset.py]: 여기도 위와 같습니다.
https://github.com/HephaestusProject/pytorch-transformers/pull/6/files#r485656553

File: src/load_dataset.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    """Load WMT14 dataset to train and test transformer
+
+    Attributes:
+        langpair: language pair to translate
+    """
+
+    def __init__(self, langpair: str) -> None:
+        super().__init__()
+        if langpair in ["de-en", "en-de", "deen", "ende"]:
+            self.dataset_config_path = dataset_config_dir / "wmt14.de-en.yaml"
```
Comment (ssaru) [inline src/load_dataset.py]: 여기도 위와 같습니다.
https://github.com/HephaestusProject/pytorch-transformers/pull/6/files#r485658524

File: src/load_dataset.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        # TODO: add en-fr
+        #  elif langpair in ["en-fr", "fr-en", "enfr", "fren"]:
+        #      self.dataset_config_path = dataset_config_dir / "wmt14.en-fr.yaml"
```
Comment (ssaru) [inline src/load_dataset.py]: `src/dataset.py`의 모듈들과 중복되는 코드가 있는 것 같습니다.
혹시 두 파일의 차이를 설명해주실 수 있을까요?

File: src/utils.py [hunk 7]
```diff
@@ -1,4 +1,14 @@
...
+    """Read text file
+
+    Args:
+        filepath: path of the test file where each line is split by '\n'
+    Returns:
+        lines: list of lines
+    """
+    with open(filepath, "r", encoding="utf-8") as f:
+        lines = [line.rstrip("\n") for line in f]
+    return lines
```
Comment (ssaru) [inline src/utils.py]: 이렇게 어떠십니까?

```suggestion
def read_lines(filepath: pathlib.Path) -> List[str]:
    """Read text file

    Args:
        filepath: path of the test file where each line is split by '\n'
    Returns:
        lines: list of lines
    """
    with filepath.open("r", encoding="utf-8") as f:
        lines = [line.rstrip("\n") for line in f]
    return lines
```

File: tests/test_dataset.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
+import pytest  # noqa: E902
```
Comment (ssaru) [inline tests/test_dataset.py:1]: 이렇게 분리해놓고 나중에 테스트 케이스를 파일로 변경해서 읽어들여서 테스트하는 방식은 어떨까요?
https://github.com/HephaestusProject/pytorch-binaryconnect/blob/master/tests/test_op_binarized_conv2d.py

File: tests/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+from pathlib import Path
+
+import pytest  # noqa: F401
+from src.utils import read_lines
+
+test_filepath = Path(__file__).parent / "data/example.de"
+
+
+@pytest.mark.parametrize("filepath", [test_filepath])
```
Comment (ssaru) [inline tests/test_utils.py]: 이렇게 어떠십니까?

```suggestion
@pytest.mark.parametrize("filepath", [Path(__file__).parent / "data/example.de"])
```

**Review by ssaru — COMMENTED** [review]
@inmoonlight 
코멘트 완료했습니다.
수고하셨습니다. 🙇‍♂️

**Comment by ssaru** [comment]
@aisolab 보섭님도 한번 봐주시지요~

File: src/dataset.py (hunk 1 above)
Comment (moon1ite) [reply src/dataset.py]: 위의 값들은 `dataset.py`의 고정 변수, 즉 외부에서 바꿀 수 있는 값이 아니어야 하므로 instance variable 이면 취지에 맞지 않는다고 생각합니다. 
대신 class variable 이면 좋을 것 같아요! 

ref: https://medium.com/python-features/class-vs-instance-variables-8d452e9abcbd

File: src/dataset.py (hunk 2 above)
Comment (moon1ite) [reply src/dataset.py]: 앗 근데 이런 방식은 다른 언어쌍이 들어올 때 대응이 안되지 않나요..? 
여기서는 보통 번역 쪽에서 langpair 입력할 때 다양한 방식이 있는데 그 처리를 해주고 싶었어요 ㅋㅋㅋ 

registry 도 한번 볼게요!

Comment (moon1ite) [reply src/dataset.py]: 올려주신 detectron dataset 은, 정말 데이터와 metadata를 등록해두고 쓰는 것 같은데 여기에는 적합한지 잘 모르겠어요 ㅋㅋ
여기서 체크하는 것은, 학습시에 언어쌍을 지정해서 사용할텐데요, 그 때 불러올 데이터셋과 토크나이저가 지정된 언어쌍에 대해서 준비된 것인지 확인하는 용도라서요! ㅎㅎ

File: src/dataset.py (hunk 1 above)
Comment (ssaru) [reply src/dataset.py]: directory를 고정하는 방식이군요!
이해했습니다.

File: src/dataset.py (hunk 2 above)
Comment (ssaru) [reply src/dataset.py]: 앗앗... 제가 NLP는 알못이라서ㅎㅎ
그냥 이런 것도 있다? 이런 느낌으로 던져보긴 했습니다. 🙇‍♂️
보섭님이 피드백 주시면 좋을 것 같긴한데.... @aisolab 호출!

File: src/dataset.py (hunk 1 above)
Comment (moon1ite) [reply src/dataset.py]: 는 configs 를 따로 관리하는 편이 코드의 가독성을 매우 높여준다고 생각해서 utils 에 `get_configs` 라는 함수를 만들었습니다! ㅎㅎ 감사해요

File: src/utils.py (hunk 7 above)
Comment (moon1ite) [reply src/utils.py]: 매우 좋습니다! ㅋㅋ

File: src/load_dataset.py (hunk 6 above)
Comment (moon1ite) [reply src/load_dataset.py]: `load_dataset` 학습에서 사용하는 train, val, test batch 를 불러오는 모듈을 담고 싶었고, `dataset` 은 불러오기 전에 tokenize 및 padding 작업을 하는 부분이었어요 
그런데 구분이 모호하고, 둘은 같은 모듈 내에 있는 것이 맞다고 생각해서 `load_dataset`을 `dataset`에 합쳤습니다! ㅎㅎ

File: tests/test_utils.py (hunk 9 above)
Comment (moon1ite) [reply tests/test_utils.py]: https://github.com/HephaestusProject/pytorch-transformers/pull/6#discussion_r485667786 과 통일성있게 테스트에 넣을 파라미터를 분리하는 방식을 고수하는 것은 어떻게 생각하세요? ㅎㅎ

File: tests/test_dataset.py (hunk 8 above)
Comment (moon1ite) [reply tests/test_dataset.py:1]: 오 매우 좋습니다! 공통적으로 들어가는 입력이라 분리하면 코드가 엄청 깔끔해지겠어요 ㅎㅎ

File: src/utils.py (hunk 7 above)
Comment (moon1ite) [reply src/utils.py]: 는 제가 filepath 를 string 으로 들고있어서 ㅋㅋ 기존의 방식대로 하겠습니다

Comment (ssaru) [reply src/utils.py]: 엌ㅋㅋ

File: tests/test_utils.py (hunk 9 above)
Comment (ssaru) [reply tests/test_utils.py]: 컼ㅋㅋㅋㅋㅋㅋㅋ 예리하십니다. 좋습니다.

**Review by ssaru — APPROVED** [review]
@inmoonlight Approve했습니다.
고생하셨습니다.

File: src/utils.py [hunk 10]
```diff
@@ -1,4 +1,79 @@
...
+        root_dir / "tokenizer" / (tokenizer_config.tokenizer_name + "-vocab.json")
+    )
+    tokenizer_config.tokenizer_merges = str(
+        root_dir / "tokenizer" / (tokenizer_config.tokenizer_name + "-merges.txt")
+    )
+
+    configs.update(
+        dataset=dataset_config, tokenizer=tokenizer_config, model=model_config
+    )
+    return configs
```
Comment (ssaru) [inline src/utils.py]: @inmoonlight 
저는 `get_config`같은 경우에는 조금 일반화해서 사용하긴 합니다.
이렇게요!

```python
def get_config(hparams: Dict, options: List) -> DictConfig:

    config: DictConfig = OmegaConf.create()

    for option in options:
        option_config: DictConfig = OmegaConf.load(hparams.get(option))
        config.update(option_config)

    OmegaConf.set_readonly(config, True)

    return config
```

File: tests/test_utils.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+test_read_lines_input = [
+    # (filepath)
+    Path(__file__).parent
+    / "data/example.de"
+]
+
+
+@pytest.mark.parametrize("filepath", test_read_lines_input)
+def test_read_lines(filepath):
+    de = read_lines(filepath)
```
Comment (ssaru) [inline tests/test_utils.py:17]: @inmoonlight 
변수명 `de`라는 것은 어떤의미일까요?

Comment (moon1ite) [reply tests/test_utils.py:17]: 독일어의 iso 표기법이에요 ㅋㅋ 

ref: https://ko.wikipedia.org/wiki/ISO_639-1_%EC%BD%94%EB%93%9C_%EB%AA%A9%EB%A1%9D

File: src/utils.py (hunk 10 above)
Comment (moon1ite) [reply src/utils.py]: 오?! 이게 훨씬 좋은 것 같아요 ㅋㅋㅋㅋㅋㅋ 반영하겠습니다 +_+ 꼼꼼히봐주셔서 감사해요 ❤️
